fixes for support of new backend

ReyhaneAskari opened the pull request:
This PR is a first attempt to fix #698, I have not added the fix for sphinx yet.

ReyhaneAskari commented:
The tests are now failing due to this error:
```AttributeError: 'module' object has no attribute 'gpuarray'```. This means the new backend is not available in the version of theano that's being used for travis. I'm going to change the check like this:

```python
if theano.config.device == 'cuda' and theano.gpuarray.pygpu_activated:
      ....
elif theano.config.device == 'gpu' and theano.sandbox.cuda.cuda_enabled:
      ....
```
Hopefully this would remove the issue. Is there a case where ```pygpu_activated``` is ```True``` but the device is not ```cuda``` (or ```cuda_enabled``` but device is not ```gpu```)? Do we want to cover them?

ReyhaneAskari commented:
The cpu tests were passed. How do you run the gpu and cuda tests?

ReyhaneAskari commented:
So I checked [the mock that you use for sphinx](https://github.com/Lasagne/Lasagne/blob/master/docs/conf.py#L319-L340). It seems that since we are supporting both backends, we will not need to change that code and it should still work.

lamblin commented on lasagne/layers/corrmm.py:
```diff
@@ -8,21 +8,26 @@
+    except:
+        raise ImportError('sandbox.cuda is not available in your\
+         version of theano.')
+else:
```
I think here, we need to also take into account the case where device == 'cpu', but either back-end has manually been activated by a call to `use(...)`.
Also, you may need to try to import theano.gpuarray (in the first case) or theano.sandbox.cuda, in the other case.

f0k commented:
> This means the new backend is not available in the version of theano that's being used for travis.

Yes, gpuarray was still in sandbox in that version. It'd be great if the check didn't depend on this, so we stay compatible to Theano 0.8.2 for the next Lasagne release (I'm not 100% sure how useful this is, but I didn't want to drop backwards compatibility too early -- I'd bump the requirement to 0.9 immediately after).

> Is there a case where pygpu_activated is True but the device is not cuda (or cuda_enabled but device is not gpu)?

Yes, see #606 for the latter case, and I guess the former case will be similar. So instead, we should try to import gpuarray from both locations. Something like:
```python
try:
    from theano import gpuarray
except ImportError:
    from theano.sandbox import gpuarray
gpu = gpuarray.pygpu_activated
if not gpu:
    try:
        from theano.sandbox import cuda
    except ImportError:
        gpu = False
    else:
        gpu = cuda.cuda_enabled
```
This would tell whether a GPU is available and activated for use.

> The cpu tests were passed. How do you run the gpu and cuda tests?

This can only be done locally, using pytest: http://lasagne.readthedocs.io/en/latest/user/development.html#testing
You will need a GPU for this.

> It seems that since we are supporting both backends, we will not need to change that code and it should still work.

Oh yes, that's true. Let's leave it as it is then, I can change it when dropping support for the old backend.

f0k commented:
To also mingle in the different imports:
```
try:
    from theano import gpuarray
except ImportError:
    from theano.sandbox import gpuarray
gpu = gpuarray.pygpu_activated
if not gpu:
    try:
        from theano.sandbox import cuda
    except ImportError:
        gpu = False
    else:
        gpu = cuda.cuda_enabled
        dnn = cuda.dnn  # or whatever module/Op is needed
else:
    dnn = gpuarray.dnn
```
After this preparation, we can through the necessary errors depending on `if gpu` and `if dnn.dnn_enabled`. (Ah, almost. `dnn_enabled` has a different name in gpuarray.)

ReyhaneAskari commented:
@f0k , thanks. I'm having problem running the tests. Could you let me know on which versions of theano this should work?

f0k commented:
> Could you let me know on which versions of theano this should work?

It should work on Theano 0.8.2, Theano 0.9.0 and the current bleeding-edge version.

> I'm having problem running the tests.

What exactly? Let me know if I can help!

ReyhaneAskari commented:
@f0k .Thanks for your reply. I don't know why part of your tests have failed. Do you suggest any fixes? I have a rather weird problem. 

Also, with Theano 0.9.X, when I set the flag to use **gpu**, this import can be done:
```from theano.sandbox import gpuarray as gpu``` but, ```gpu.pygpu_activated``` is false. I manually double checked ```theano.config.device``` and it was gpu. Do you have any idea why pygpu_activated doesn't work?

nouiz commented:
the Theano flag set to "gpu" mean that it is sandbox/cuda that is used. So
it is normal that pygpu (the new back-end) is not activated.

On Wed, May 17, 2017 at 5:51 PM Reyhane Askari <notifications@github.com>
wrote:

> @f0k <https://github.com/f0k> .Thanks for your reply. I don't know why
> part of your tests have failed. Do you suggest any fixes? I have a rather
> weird problem.
>
> Also, with Theano 0.9.X, when I set the flag to use *gpu*, this import
> can be done:
> from theano.sandbox import gpuarray as gpu but, gpu.pygpu_activated is
> false. I manually double checked theano.config.device and it was gpu. Do
> you have any idea why pygpu_activated doesn't work?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Lasagne/Lasagne/pull/836#issuecomment-302241926>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALC-0wASBlMsfbFM5Uy4sy1JDFqqktVks5r62vXgaJpZM4NarF1>
> .
>

f0k commented:
> I don't know why part of your tests have failed.

The tests were fine, the coverage was decreased. As before, the parts that require a GPU must be marked as `# pragma: no cover`.

And thanks for separating out the GPU check, but we cannot place it in `lasagne.__init__`. This means that when somebody uses Theano 0.9 with a CPU, an `import lasagne` will import both `gpuarray` and `sandbox.cuda`, which a) slows things down unnecessarily and b) triggers a warning that `sandbox.cuda` is deprecated. These imports must only happen if somebody explicitly tries to import `lasagne.layers.dnn` or `lasagne.layers.corrmm`.
As it only affects two files, I'd suggest to just take the snippet I posted in https://github.com/Lasagne/Lasagne/pull/836#issuecomment-301554552 and replicate it -- once including the dnn check, once without. Can you do this?

f0k commented on lasagne/layers/dnn.py:
```diff
@@ -1,6 +1,4 @@
 import theano
-from theano.sandbox.cuda import dnn
-
```
The empty line was on purpose; PEP 8 suggests to group imports into standard library, third-party, and internal (https://www.python.org/dev/peps/pep-0008/#imports).

ReyhaneAskari replied:
ok, I will revert it.

f0k commented on lasagne/layers/dnn.py:
```diff
@@ -10,15 +8,27 @@
+else:
+    raise ImportError("requires GPU support -- see http://lasagne.\
+                readthedocs.org/en/latest/user/installation.html#gpu-support")
+if not dnn_available:
```
You sometimes call it `dnn_avail` and sometimes `dnn_available`.

ReyhaneAskari replied:
thank, done.

f0k commented on lasagne/layers/dnn.py:
```diff
@@ -10,15 +8,27 @@
-            "latest/user/installation.html#cudnn" %
-            dnn.dnn_available.msg)  # pragma: no cover
+        "cuDNN not available: %s\nSee http://lasagne.readthedocs.org\
+        /en/latest/user/installation.html#cudnn")  # pragma: no cover
```
You removed the helpful message Theano puts into `dnn_available.msg` when the import failed, but you left in the `%s`.

ReyhaneAskari replied:
I wasn't sure that there is such message for ```dnn_present()```. I will double check and try to keep the message.

ReyhaneAskari commented:
@f0k , thanks. I didn't know about ```# pragma: no cover```.  I will change the code such that those checks will be local. I made it central because specially in the case where we checked the dnn it was hard to walk through.

ReyhaneAskari commented:
I ran the tests on the gpu and there are two failing tests that I will work on them tomorrow. It is ```lasagne/tests/layers/test_special.py:729: AssertionError``` and ```lasagne/tests/layers/test_noise.py:141: AttributeError```. The first one is failing when the device is ```gpu``` and ```cuda``` the second one only fails when the device is ```cuda```.

ReyhaneAskari commented on lasagne/layers/dnn.py:
```diff
@@ -31,8 +32,7 @@
             "cuDNN not available: %s\nSee http://lasagne.readthedocs.org\
-            /en/latest/user/installation.html#cudnn"
-            % dnn.dnn_available.msg)  # pragma: no cover
+            /en/latest/user/installation.html#cudnn")  # pragma: no cover
```
I removed it for now so that the tests would pass and fix it later.

f0k commented on lasagne/tests/layers/test_noise.py:
```diff
@@ -138,7 +138,7 @@ def test_get_output_for_deterministic(self, layer):
         result = layer.get_output_for(input, deterministic=True)
         result_eval = result.eval()
-        assert (result_eval == input.eval()).all()
+        assert (np.array(result_eval) == np.array(input.eval())).all()
```
Ah... so in the new backend, `eval()` does not always return a ndarray?

ReyhaneAskari replied:
It returned a ```gpuarray ndarray``` and I think the problem is that the equal method is not defined properly. So I had to convert it to an ```np.array```.

f0k replied:
I see, it returns a `GpuArray` for which `==` only checks for identity. A shorter variant would be `np.allclose(result.eval(), input.eval())`, which handles the conversion. The tests above will need to be changed as well then -- they assert that the comparison is false, but if the comparison only checks for identity, this is trivially fulfilled.

Also for some reason, the input is not floatX, I think this was not intended. It's not terribly important for the test; we can rely on Theano that CPU and GPU versions are the same (we still might want to run the test suite with float64 warnings enabled and see if there are more glitches like this). But why is the result still a gpuarray then? Ah, I see, it's a float64 GpuArray!

Phew. I think we may need to go through all the tests and see if they're using `==` or `!=` where they should use `np.allclose` instead.

ReyhaneAskari replied:
I did actually check the ```np.allclose``` and it was true. So I was really surprised at first that the ```==``` gave ```false```.  I think changing all the tests could be another PR but for now I will change this one to ```np.allclose``` and test it with float64 and float32.

f0k replied:
> I think changing all the tests could be another PR

Yes, this doesn't have to happen on the spot.

> I will change this one to `np.allclose`

:+1:

> and test it with float64 and float32

Please only test with floatX -- we can run the suite with different floatX settings if we want to.

ReyhaneAskari replied:
Thanks, yes that's what I meant. I will use floatX in the code and test it with the theano flag of floatX=float32 and floatX=falot64.

f0k commented on lasagne/tests/layers/test_special.py:
```diff
@@ -633,8 +633,8 @@ def ParametricRectifierLayer(self):
-        return lambda shape: (np.arange(np.prod(shape)).reshape(shape)) \
-            / np.prod(shape)
+        return lambda shape: ((np.arange(np.prod(shape)).reshape(shape)) /
+                              np.prod(shape)).astype(theano.config.floatX)
```
Uh! I hadn't noticed when reviewing #446. This is wrong -- it will always return an array of zeros in Python2. We could add a `__future__` import, but I'd prefer to fix this as:
```python
from lasagne.utils import floatX
return lambda shp: (floatX(np.arange(np.prod(shp)).reshape(shp)) /
                      floatX(np.prod(shp)))
```
In theory, I should do a separate PR for this, but then you'd need to undo your change and rebase -- can you just edit it with my fix?

ReyhaneAskari replied:
Yes, I made this change but still there seemed to be a small difference in the values for some tests. So I changed the ```atol``` of ```np.allclose``` from the default value of 1e-08 to 1e-07 in two cases.

ReyhaneAskari commented:
I saw that you had two imports for numpy in this file: lasagne/tests/layers/test_noise.py
So I removed one of them and converted all the numpys to np.

ReyhaneAskari commented:
I don't know why the test for coverage is failed and what should be done. I will wait for @f0k to review the PR completely and then I will run the tests for device = **cpu**, **gpu**, **cuda** with **float32** and **float64** using Theano **dev**, **0.8** and **0.9** versions.

ReyhaneAskari commented:
@nouiz I created an issue here: https://github.com/Theano/libgpuarray/issues/443

f0k commented:
> So I removed one of them and converted all the numpys to np.

Okay, thanks... many tests have `import numpy` for some reason, this should probably be cleaned up in a separate PR.

> I don't know why the test for coverage is failed and what should be done.

Hmm, this is indeed strange. It says https://coveralls.io/builds/11709461/source?filename=lasagne%2Futils.py#L371 is not covered although it was covered before. It must have been accidentally covered by one of the tests that you changed, maybe via the `alpha` definition. It should instead be covered by a test in https://github.com/Lasagne/Lasagne/blob/master/lasagne/tests/test_utils.py.
Probably the cleanest solution would be to add a test in another PR and rebase this one. Or add a commit here.

Sorry for all the unrelated problems this triggered, and sorry for being so slow with reviewing, I'm writing my thesis. If you don't mind, I can try to take over this PR, rearrange some commits, add the missing test -- this might be easier than going back and forth all the time. Let me know if that's okay with you. Otherwise, the next step would be to add a test that covers the line that's currently missing.

ReyhaneAskari commented:
No problem at all. There is no rush. I created a PR for the missing tests. I will rebase after that one got merged. Regarding the numpy, I can get rid of that commit if you'd prefer and create a separate PR for it.

f0k commented:
> I created a PR for the missing tests.

Great, thank you!

> Regarding the numpy, I can get rid of that commit if you'd prefer and create a separate PR for it.

Yes, that'd be better, so that the history is easier to understand afterwards. It'd be best to address all `import numpy` at once then. We can leave that for later, or just create an issue for now, it's not a pressing matter.

ReyhaneAskari commented:
@f0k , so I rebased and removed the commit regarding to numpy and squashed some other commits. Please let me know if the rest is good. Then I will test it with different device and different versions.

f0k commented on lasagne/layers/dnn.py:
```diff
@@ -10,16 +9,30 @@
+if not dnn_avail:
+        raise ImportError(
+            "cuDNN not available:See http://lasagne.readthedocs.org\
+            /en/latest/user/installation.html#cudnn")  # pragma: no cover
```
You're still missing the message here. In line with the suggestion above, I would suggest:
```python
# check if Theano's new GPU backend is available and in use
try:
    from theano import gpuarray as gpu
except ImportError:
    from theano.sandbox import gpuarray as gpu
gpu_enabled = gpu.pygpu_activated
dnn_enabled = gpu.dnn.dnn_present
# if not, try to fall back to Theano's old GPU backend
if not gpu_enabled:
    try:
        from theano.sandbox import cuda as gpu
    except ImportError:
        gpu_enabled = False  # [I know it's False already but it is easier to follow]
    else:
        gpu_enabled = gpu.cuda_enabled
        dnn_enabled = gpu.dnn.dnn_available
# if either of the backends is available, use it, otherwise bail out
if gpu_enabled:
    if dnn_enabled():
        dnn = gpu.dnn
    else:
        raise ImportError("cuDNN not ..." % dnn_enabled.msg)
else:
    raise ImportError("requires GPU ...")
```

ReyhaneAskari replied:
I will change this one as well.

f0k commented on lasagne/tests/layers/test_conv.py:
```diff
@@ -7,6 +7,25 @@
 import lasagne
 from lasagne.utils import floatX, as_tuple
 
+theano_backend = "cpu"
```
You can remove this line.

f0k commented on lasagne/tests/layers/test_conv.py:
```diff
@@ -7,6 +7,25 @@
+    theano_backend = "pygpu_sandbox"
+gpu = gpuarray.pygpu_activated
+if not gpu:
+    theano_backend = "cpu"
```
You can remove this line.

ReyhaneAskari commented:
@f0k , made all your requested changes.

ReyhaneAskari commented on lasagne/layers/corrmm.py:
```diff
@@ -8,21 +8,30 @@
+            "latest/user/installation.html#gpu-support")  # pragma: no cover
+if enabled:
+    from gpu.basic_ops import gpu_contiguous
+    from gpu.blas import GpuCorrMM
```
Yes this is nicer.

ReyhaneAskari commented:
I don't get these errors locally when I run ``` py.test -q -s```. Do you have any idea?

lamblin commented:
This is with Theano 0.8.0, with no GPU back-end present.
Maybe `theano.sandbox.cuda.dnn` needs to be explicitly imported, rather than just importing `theano.sandbox.cuda`. In fact, the [original code](https://github.com/Lasagne/Lasagne/pull/836/files#diff-185d143ba129327c0ce74dd75858134aL2) did that.
The new back-end seems to automatically import `.dnn` when we import `theano.gpuarray` or `theano.sandbox.gpuarray`.

f0k commented on lasagne/layers/corrmm.py:
```diff
@@ -8,21 +8,31 @@
+else:
+    raise ImportError(
+        "requires GPU support -- see http://lasagne.readthedocs.org/en/"
+        "latest/user/installation.html#gpu-support")  # pragma: no cover
```
Can you please also include the comments I had in my suggestion? (Except for the one in square brackets.)

f0k commented on lasagne/layers/dnn.py:
```diff
@@ -10,16 +9,32 @@
+dnn_enabled = gpu.dnn.dnn_present
+if not gpu_enabled:
+    try:
+        from theano.sandbox import cuda as gpu
```
As per @lamblin's comment, we need to explicitly import dnn for the old backend. The simplest fix would be to add a line here (before the `except`): `import theano.sandbox.cuda.dnn`. This will also make dnn available as `gpu.dnn`.

A cleaner version would be to assign `dnn = gpu.dnn` above, after `dnn_enabled = ...`, and add `from theano.sandbox.cuda import dnn` here, and change the `if/else` construct below, but I'd be fine with the simpler solution.

f0k commented on lasagne/layers/dnn.py:
```diff
@@ -10,16 +9,32 @@
-            dnn.dnn_available.msg)  # pragma: no cover
-
+        "requires GPU support -- see http://lasagne.readthedocs.org/en/"
+        "latest/user/installation.html#gpu-support")  # pragma: no cover
```
Again, please include the comments I had in my suggestion.

f0k commented on lasagne/layers/dnn.py:
```diff
@@ -10,16 +9,32 @@
+        raise ImportError(
+            "cuDNN not available:See http://lasagne.readthedocs.org\
+            /en/latest/user/installation.html#cudnn\
+            " % dnn_enabled.msg)  # pragma: no cover
```
Be careful, you included the `% dnn_enabled.msg` but not the `%s`.

ReyhaneAskari commented:
@lamblin thanks. I did actually check the other versions but I still didn't have that error. I made the change suggested by @f0k , hopefully it will fix it.

ReyhaneAskari commented:
@f0k , thanks. Unfortunately I don't have more time for this ticket this week. There is a failing test for the coverage. Also, I think we didn't really need to change the order of the commits. People usually update their repo with a merge commit and every commit doesn't need to be a working version on a dev branch. I had to resolve 3 conflicts to change the order of these two commits.

f0k commented:
> Also, I think we didn't really need to change the order of the commits. People usually update their repo with a merge commit and every commit doesn't need to be a working version on a dev branch.

Having a clean history is not needed for updating -- otherwise we could also leave all the separate fixes in there. It's to trace back potential problems later. E.g., with a clean history, you can use `git blame` to track the history of a particular line and understand what context it was introduced or changed for. And you can `git bisect` a problem and always end up in valid states of the code.

> I had to resolve 3 conflicts to change the order of these two commits.

Sorry, something must have gone wrong then. The two commits change different files, so it should be possible to change their order freely with a `git rebase -i`. Otherwise I wouldn't have suggested it. I may be pedantic, but not at all costs :)

> There is a failing test for the coverage.

This is weird, this means the code has changed. I'll look into it. Thanks for all the work!

f0k commented:
> I'll look into it.

Interesting. There's a line that's only covered in Python 2. Usually coveralls compiles its report from both the Python 2 and Python 3 run: https://coveralls.io/builds/11789199
In this case, it only caught the Python 3 run: https://coveralls.io/builds/11787955
I'll restart Travis and hope this will fix it. Besides, I'll check if there's a way to test for that line in Python 3.

f0k commented:
> I'll restart Travis and hope this will fix it.

It did!

> Besides, I'll check if there's a way to test for that line in Python 3.

There isn't. We can either exclude it from coverage completely, write a different .coveragerc for Python 3, or rely on coveralls to merge the Python 2 and 3 reports as it should. I guess I'll leave it as it is for now.

I'll merge this PR in the evening if there are no further comments.

ReyhaneAskari commented:
@f0k , perfect! Thanks a lot!

ReyhaneAskari commented:
@f0k , I wanted to test this with **device = cpu, gpu, cuda** and **floatX=float32,float64** and theano versions o**f 0.8.X and 0.9.X and master.** I'll do the tests in the weekend and update you.

f0k commented:
> I'll do the tests in the weekend and update you.

Okay, thanks! I'll wait with merging then, just in case. When testing 0.8.x, use 0.8.2 if in doubt. The requirement is currently set to 0.8.0, but I'll update it for releasing Lasagne 0.2.

ReyhaneAskari commented:
Hey @f0k , so I tested everything and removed the ```.pyc``` and theano cache after each test :

1. On theano/master, with ```device=gpu```, we have this error which is expected: ```ValueError: You are tring to use the old GPU back-end. It was removed from Theano. Use device=cuda* now. ```
2. On theano/0.8.X with ```device=gpu``` and ```floatX=float64```, we have many errors. I don't know if it was working before and these are expected. It is attached [here](https://github.com/Lasagne/Lasagne/files/1052332/8_gpu_64.txt). It works fine with ```float32```.

The rest were all fine.

nouiz commented:
The old back-end (device=gpu) only support float32. So testing it with
float64 isn't a good think to do, as many things will fail. So it seem all
is good.

On Mon, Jun 5, 2017 at 11:19 AM Reyhane Askari <notifications@github.com>
wrote:

> Hey @f0k <https://github.com/f0k> , so I tested everything and removed
> the .pyc and theano cache after each test :
>
>    1. On theano/master, with device=gpu, we have this error which is
>    expected: ValueError: You are tring to use the old GPU back-end. It
>    was removed from Theano. Use device=cuda* now.
>    2. On theano/0.8.X with device=gpu and floatX=float64, we have many
>    errors. I don't know if it was working before and these are expected. It is
>    attached here
>    <https://github.com/Lasagne/Lasagne/files/1052332/8_gpu_64.txt>. It
>    works fine with float32.
>
> The rest were all fine.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Lasagne/Lasagne/pull/836#issuecomment-306215860>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALC--iHTe_m9fpj0qYixz50Dhh5oBTBks5sBByYgaJpZM4NarF1>
> .
>

f0k commented:
Thank you for the extensive test! So the errors you had with float64 and device=gpu were all `TypeError: CudaNdarrayType only supports dtype float32 for now. Tried using dtype float64 for variable None`. This is expected.

Merging this, thanks again!

ReyhaneAskari commented:
Hi @f0k , we are willing to contribute more to Lasagne, specially for the issues that would help the release of Lasagne so that the old backend of theano won't be used anymore. Do you have any suggestions where I can help you?

f0k commented:
> we are willing to contribute more to Lasagne, specially for the issues that would help the release of Lasagne

That's great! I'll also be able to contribute more myself when I've finished my thesis (probably in a month), but help for releasing 0.2 would be greatly appreciated. It's taking much longer than anticipated.

> Do you have any suggestions where I can help you?

In general, the issues/PRs to be addressed for 0.2 are listed in this milestone:
https://github.com/Lasagne/Lasagne/milestone/2
I just went through the list and moved some to 0.3 that would just be nice to have, but are not essential. Also I tagged all with "documentation" that only require documentation changes, and one with "easy".
Most of the issues are relatively straightforward. Two are meant to straighten the interface before setting it in stone: #719, #399. Those will require some thought/discussion.
Feel free to browse the list and take anything you want -- just leave a post that you're working on it, and please ask if anything is unclear. Looking forward to your contributions!

ReyhaneAskari commented:
Thanks @f0k for your explanations. I will start working on https://github.com/Lasagne/Lasagne/issues/706 tomorrow or next week.

f0k commented:
Great! For the future: with "just leave a post" I meant "leave a post at the corresponding issue that you plan to work on" :) Just so nobody else takes it (which is unlikely given past experience, but anyway).
